### PR TITLE
script: Fix panic in `htmlimageelement.rs` using `str::find()` to find character boundaries.

### LINKED ISSUE #32976: Panic parsing srcset in html/semantics/embedded-content/the-img-element/srcset/parse-a-srcset-attribute.html
To reproduce: `./mach test-wpt tests/wpt/tests/html/semantics/embedded-content/the-img-element/srcset/parse-a-srcset-attribute.html`

The code in https://github.com/servo/servo/blob/main/components/script/dom/htmlimageelement.rs#L1894-L1897 is misusing the chars iterator to obtain byte indexes from the string. We should use the [char_indices](https://doc.rust-lang.org/stable/std/primitive.str.html#method.char_indices) iterator instead.

Backtrace:
```
pid:15614 byte index 1 is not a char boundary; it is inside '\u{a0}' (bytes 0..2) of ` data:,a` (thread Script(1,1), at components/script/dom/htmlimageelement.rs:1896)
 0:07.14 pid:15614   14: core::str::traits::<impl core::slice::index::SliceIndex<str> for core::ops::range::Range<usize>>::index
 0:07.14 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/str/traits.rs:241:21
 0:07.14 pid:15614   15: core::str::traits::<impl core::ops::index::Index<I> for str>::index
 0:07.14 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/str/traits.rs:61:9
 0:07.18 pid:15614   16: script::dom::htmlimageelement::collect_sequence_characters
 0:07.18 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:1896:23
 0:07.22 pid:15614   17: script::dom::htmlimageelement::parse_a_srcset_attribute
 0:07.22 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:1911:13
 0:07.22 pid:15614   18: script::dom::htmlimageelement::HTMLImageElement::update_source_set
 0:07.22 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:586:48
 0:07.22 pid:15614   19: script::dom::htmlimageelement::HTMLImageElement::select_image_source
 0:07.22 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:745:9
 0:07.22 pid:15614   20: script::dom::htmlimageelement::HTMLImageElement::update_the_image_data_sync_steps
 0:07.22 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:871:42
 0:07.22 pid:15614   21: <script::dom::htmlimageelement::ImageElementMicrotask as script::microtask::MicrotaskRunnable>::handler
 0:07.22 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:1427:21
 0:07.26 pid:15614   22: script::microtask::MicrotaskQueue::checkpoint
 0:07.26 pid:15614              at /Users/jdm/src/servo/components/script/microtask.rs:134:25
 0:07.29 pid:15614   23: script::script_thread::ScriptThread::perform_a_microtask_checkpoint
 0:07.29 pid:15614              at /Users/jdm/src/servo/components/script/script_thread.rs:4159:13
 0:07.29 pid:15614   24: script::script_thread::ScriptThread::handle_msgs
 0:07.29 pid:15614              at /Users/jdm/src/servo/components/script/script_thread.rs:1973:13
 0:07.29 pid:15614   25: script::script_thread::ScriptThread::start
 0:07.29 pid:15614              at /Users/jdm/src/servo/components/script/script_thread.rs:1454:15
 0:07.32 pid:15614   26: <script::script_thread::ScriptThread as script_layout_interface::ScriptThreadFactory>::create::{{closure}}::{{closure}}
 0:07.32 pid:15614              at /Users/jdm/src/servo/components/script/script_thread.rs:845:25
 0:07.35 pid:15614   27: profile_traits::mem::ProfilerChan::run_with_memory_reporting
 0:07.35 pid:15614              at /Users/jdm/src/servo/components/shared/profile/mem.rs:91:9
 0:07.35 pid:15614   28: <script::script_thread::ScriptThread as script_layout_interface::ScriptThreadFactory>::create::{{closure}}
 0:07.35 pid:15614              at /Users/jdm/src/servo/components/script/script_thread.rs:843:17
 0:07.35 pid:15614   29: std::sys_common::backtrace::__rust_begin_short_backtrace
 0:07.35 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/sys_common/backtrace.rs:155:18
 0:07.38 pid:15614   30: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
 0:07.38 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/thread/mod.rs:528:17
 0:07.41 pid:15614   31: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
 0:07.41 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panic/unwind_safe.rs:272:9
 0:07.44 pid:15614   32: std::panicking::try::do_call
 0:07.44 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:552:40
 0:07.44 pid:15614   33: ___rust_try
 0:07.44 pid:15614   34: std::panicking::try
 0:07.44 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:516:19
 0:07.44 pid:15614   35: std::panic::catch_unwind
 0:07.44 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panic.rs:146:14
 0:07.44 pid:15614       std::thread::Builder::spawn_unchecked_::{{closure}}
 0:07.44 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/thread/mod.rs:527:30
 0:07.82 pid:15614   36: core::ops::function::FnOnce::call_once{{vtable.shim}}
 0:07.82 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/ops/function.rs:250:5
 0:07.82 pid:15614   37: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
 0:07.82 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/boxed.rs:2020:9
 0:07.82 pid:15614       <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
 0:07.82 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/boxed.rs:2020:9
 0:07.82 pid:15614       std::sys::pal::unix::thread::Thread::new::thread_start
 0:07.82 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/sys/pal/unix/thread.rs:108:17
 0:07.82 pid:15614   38: __pthread_joiner_wake
```

### PR DESCRIPTION
**Pull request by NotnaKO**
In function collect_sequence_characters replace loop with .chars().enumerate() by .find and split with its index.
In function parse_a_srcset_attribute many fixes in algo https://html.spec.whatwg.org/multipage/images.html#parsing-a-srcset-attribute:

1) replace checking whitespace to checking ascii_whitespaces according to the algo
2) check more carefully if in 8
3) replace buffered variable to peekable iterator
4) drop chars().enumerate() in descriptor's loop in 8
5) rewrite loop 13: fix parsing digits and remaining(w|x|h), add checks in guards that whole numbers are parsed (because functions parse_unsigned_integer and parse_double parse only valid prefix of input)

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #32976 

- [X] There are tests for these changes OR

### CODE REVIEW

**Comment by mrobinson** [comment]
This has the same failures that showed up in #32978, so maybe we can close this and discuss fixing them there?

**Comment by NotnaKO** [comment]
@mrobinson I think now it fix parse_a_srcset_attribute

**Comment by mrobinson** [comment]
@NotnaKO Thanks. Can you describe the changes you made in the PR body as well as the adaptations you had to make in `parse_a_srcset_attribute`? This will help make it easier to review your change and will make it more useful for people looking at the commit history.

**Comment by NotnaKO** [comment]
> @NotnaKO Thanks. Can you describe the changes you made in the PR body as well as the adaptations you had to make in `parse_a_srcset_attribute`? This will help make it easier to review your change and will make it more useful for people looking at the commit history.

Now better?

**Comment by mrobinson** [comment]
> > @NotnaKO Thanks. Can you describe the changes you made in the PR body as well as the adaptations you had to make in `parse_a_srcset_attribute`? This will help make it easier to review your change and will make it more useful for people looking at the commit history.
> 
> Now better?

Yes, thanks. This is helpful.

**Comment by NotnaKO** [comment]
@mrobinson Can you add the label to the tests again, please? (I didn't know about ./mach fmt, but I fixed it now, sorry for that)

**Comment by Taym95** [comment]
> @mrobinson Can you add the label to the tests again, please? (I didn't know about ./mach fmt, but I fixed it now, sorry for that)

easy way if to run this label/workflow on your fork

**Comment by NotnaKO** [comment]
> Test results for linux-wpt-layout-2020 from try job ([#10350103714](https://github.com/servo/servo/actions/runs/10350103714)):
> 
> Flaky unexpected result (19)
> Stable unexpected results that are known to be intermittent (12)
> Stable unexpected results (7)
> * OK `/html/semantics/embedded-content/the-img-element/current-pixel-density/basic.html`* FAIL [expected PASS] subtest: `&lt;img srcset="/images/green-256x256.png 9e99999999999999999999999x" data-expect="0"&gt;`
>     ```
>     assert_equals: width expected 0 but got 28
>     ```
> * PASS [expected FAIL] `/html/semantics/embedded-content/the-img-element/sizes/sizes-auto-rendering-2.html`
> * FAIL [expected PASS] `/html/semantics/embedded-content/the-img-element/sizes/sizes-auto-rendering-3.html`
> * FAIL [expected PASS] `/html/semantics/embedded-content/the-img-element/sizes/sizes-auto-rendering.html`
> * OK [expected CRASH] `/html/semantics/embedded-content/the-img-element/srcset/parse-a-srcset-attribute.html`
> * OK `/html/semantics/embedded-content/the-img-element/update-the-image-data/fail-to-resolve.html`* PASS [expected FAIL] subtest: `&lt;img srcset="//[, /images/red.png"&gt;`
> * OK `/html/semantics/embedded-content/the-img-element/update-the-source-set.html`* PASS [expected FAIL] subtest: `&lt;img src="data:,a" srcset="data:,b 1e0x" data-expect="data:,b"&gt;`

I didn't quite understand, those tests that fell, although it was expected to pass, is this a problem or are the tests like this? Auto-rendering tests write that 429 pixels have changed, how can I understand something is wrong or not? And another test basic.html, which from 9e9999999999999999999999x probably got 0, because there was a number parsing error in the previous version of the algorithm (it seems that 9 was taken there, then there was e, not x, which made an error).

**Comment by mrobinson** [comment]
> I didn't quite understand, those tests that fell, although it was expected to pass, is this a problem or are the tests like this? Auto-rendering tests write that 429 pixels have changed, how can I understand something is wrong or not? And another test basic.html, which from 9e9999999999999999999999x probably got 0, because there was a number parsing error in the previous version of the algorithm (it seems that 9 was taken there, then there was e, not x, which made an error).

A test failure typically indicates a bug in the implementation. The first step is to understand what part of the specification the test is trying to test and the second step is to figure out why our implementation doesn't match up with what the specification says. In very rare cases the expected behavior does not match up with what the specification says, but this is very rare.

**Comment by NotnaKO** [comment]
Ok, I check the test auto-rendering.html and now I more sure that old version is wrong and probably test too. In test we have:
```js
testImg.setAttribute('srcset', `
    /images/fail.gif 10w,
    /images/green.png 100w,
    /images/red.png 1000w
  `);
```
So we have a call of the function `parse_a_srcset_attribute` with input 
```rs
"
                /images/fail.gif 10w,
                /images/green.png 100w,
                /images/red.png 1000w
"
```
, right?
I checked descriptors in the old version and they are 
```
["10w"]
["100w"]
["1000w\n"]
```
Here you can see the last \n. It is here, because in old version we check just space (although in html standard we have ascii_whitespace, which is not just space). In other descriptors we don't have \n, because we have a comma after them. So the remaining is not equal  "w" and we have an error in last descriptor.
That's why result of the old version of function is  
```rs
[ImageSource { url: "/images/fail.gif", descriptor: Descriptor { wid: Some(10), den: None } },
 ImageSource { url: "/images/green.png", descriptor: Descriptor { wid: Some(100), den: None } }]
```
And new version returns 
```rs
[ImageSource { url: "/images/fail.gif", descriptor: Descriptor { wid: Some(10), den: None } }, 
ImageSource { url: "/images/green.png", descriptor: Descriptor { wid: Some(100), den: None } }, 
ImageSource { url: "/images/red.png", descriptor: Descriptor { wid: Some(1000), den: None } }]
```
That's why I have different 429 pixels in test.

**Comment by jdm** [comment]
Oh, to be clear: when it says "expected FAIL", it means that Servo's current stable test result is a failure (for the reasons that you described). In that case, a PASS expected FAIL result is a good outcome, and means we can update the stable results to reflect that Servo us now expected to pass the test. https://book.servo.org/hacking/testing.html#updating-web-test-expectations

**Comment by jdm** [comment]
But yeah, a new FAIL expected PASS result is surprising. I misread which tests you were investigating.

File: components/script/dom/htmlimageelement.rs [hunk 1]
```diff
@@ -1891,164 +1891,143 @@ pub fn collect_sequence_characters<F>(s: &str, predicate: F) -> (&str, &str)
...
-            }
-            match state {
-                ParseState::InDescriptor => match next_char {
-                    Some((_, ' ')) => {
-                        if !current_descriptor.is_empty() {
-                            descriptors.push(current_descriptor.clone());
-                            current_descriptor = String::new();
-                            state = ParseState::AfterDescriptor;
-                        }
-                        continue;
```
Comment (NotnaKO) [inline components/script/dom/htmlimageelement.rs:1949]: I think this is the bug with '/n'. We need char::is_ascii_whitespace check like in new version.

**Comment by NotnaKO** [comment]
I also discover test basic.html, and find a very strange behaivor in parse_double function:
I see I call parse_double(9e99999999999999999999999) and result is Ok(9.0). This is a bug in function read_exponent (in style::str) in parse_double in line:
```rs
read_numbers(iter).0.map(|exp| exp.to_i32().unwrap_or(0))
```
We read 99999999999999999999999 and it doesn't convert to i32, so return 0 and we have Ok(9.0).

**Comment by mrobinson** [comment]
I have pushed a new commit to this branch that:

1. Reworks a bit of the algorithm that you have written and added specification language to make it more readable. I needed to do this to understand that the code was following the specification and what it was doing.
2. Switched the double parsing to use Rust's builtin parser. The HTML specification has different requirements for parsing than the CSS specification. I've added a comment giving more details about the situation -- essentially we need to filter strings before we parse them like Gecko does in https://searchfox.org/mozilla-central/rev/a85b25946f7f8eebf466bd7ad821b82b68a9231f/dom/base/ResponsiveImageSelector.cpp#36.
3. Updated test expectations. The two new failures are due to the fact that we do not support the `auto` value for the `sizes` attribute. Being better at parsing `srcset` uncovered these failures.

**Review by mrobinson — APPROVED** [review]

**Comment by mrobinson** [comment]
It looks like one of the pre-existing unit tests failed because it expected an invalid `srcset` string to be parsed. I've updated the test and pushed to the branch again.

**Comment by jdm** [comment]
```
 ▶ OK /html/semantics/embedded-content/the-img-element/srcset/select-an-image-source.html
  │   ▶ FAIL [expected PASS] "data:,a 1w 1h, data:,b 1w"
  │   │   → assert_equals: expected "data:,a" but got "data:,b"
  │   │ FAIL [expected PASS] "data:,a 1w 1h, data:,b 1w 2h"
  │   │   → assert_equals: expected "data:,a" but got ""
  │   │ FAIL [expected PASS] "data:,a 1w 2h, data:,b 1w 1h"
  │   │   → assert_equals: expected "data:,a" but got ""
  │   │ 
  │   │ check/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:18:18
  │   │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2622:25
  │   │ test@http://web-platform.test:8000/resources/testharness.js:633:30
  │   │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:11:7
  └   └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:23:14
```

**Comment by mrobinson** [comment]
> ```
> 
>  ▶ OK /html/semantics/embedded-content/the-img-element/srcset/select-an-image-source.html
> 
>   │   ▶ FAIL [expected PASS] "data:,a 1w 1h, data:,b 1w"
> 
>   │   │   → assert_equals: expected "data:,a" but got "data:,b"
> 
>   │   │ FAIL [expected PASS] "data:,a 1w 1h, data:,b 1w 2h"
> 
>   │   │   → assert_equals: expected "data:,a" but got ""
> 
>   │   │ FAIL [expected PASS] "data:,a 1w 2h, data:,b 1w 1h"
> 
>   │   │   → assert_equals: expected "data:,a" but got ""
> 
>   │   │ 
> 
>   │   │ check/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:18:18
> 
>   │   │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2622:25
> 
>   │   │ test@http://web-platform.test:8000/resources/testharness.js:633:30
> 
>   │   │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:11:7
> 
>   └   └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:23:14
> 
> ```

Hrm. Looks like there's a bit more work to do here.

**Pull request merged**